graphql_compiler/query_pagination/query_parameterizer.py

```diff
@@ -1,61 +1,288 @@
-        # for pagination in the page query.
-        "remainder_query_filter",  # Directive, filter directive with '>=' operator usable
-        # for pagination in the remainder query.
-        "related_filters",  # List[Directive], filter directives that share the same
-        # vertex and property field as the next_page_query_filter,
-        # and are used to generate more accurate pages.
-    ),
+from graphql.language.ast import (
+    ArgumentNode,
+    DirectiveNode,
+    DocumentNode,
+    FieldNode,
+    InlineFragmentNode,
+    ListValueNode,
+    NameNode,
+    OperationDefinitionNode,
+    SelectionSetNode,
+    StringValueNode,
 )
 
+from ..ast_manipulation import get_ast_field_name, get_only_query_definition
+from ..compiler.helpers import get_parameter_name
+from ..exceptions import GraphQLError
+from ..global_utils import ASTWithParameters
+from ..schema.schema_info import QueryPlanningSchemaInfo
+from .pagination_planning import VertexPartitionPlan
 
-def generate_parameterized_queries(schema_info, query_ast, parameters):
-    """Generate two parameterized queries that can be used to paginate over a given query.
 
-    In order to paginate arbitrary GraphQL queries, additional filters may need to be added to be
-    able to limit the number of results in the original query. This function creates two new queries
-    with additional filters stored as PaginationFilters with which the query result size can be
-    controlled.
+def _generate_new_name(base_name: str, taken_names: Set[str]) -> str:
+    """Return a name based on the provided string that is not already taken.
+
+    This method tries the following names: {base_name}_0, then {base_name}_1, etc.
+    and returns the first one that's not in taken_names
+
+    Args:
+        base_name: The base for name construction as explained above
+        taken_names: The set of names not permitted as output
+
+    Returns:
+        a name based on the base_name that is not in taken_names
+    """
+    index = 0
+    while "{}_{}".format(base_name, index) in taken_names:
+        index += 1
+    return "{}_{}".format(base_name, index)
+
+
+def _get_binary_filter_node_parameter(filter_directive: DirectiveNode) -> str:
+    """Return the parameter name for a binary Filter Directive."""
+    filter_arguments = cast(ListValueNode, filter_directive.arguments[1].value).values
+    if len(filter_arguments) != 1:
+        raise AssertionError(f"Expected one argument in filter {filter_directive}")
+
+    argument_name = cast(StringValueNode, filter_arguments[0]).value
+    parameter_name = get_parameter_name(argument_name)
+    return parameter_name
+
+
+def _get_filter_node_operation(filter_directive: DirectiveNode) -> str:
+    """Return the @filter's op_name as a string."""
+    return cast(StringValueNode, filter_directive.arguments[0].value).value
+
+
+def _is_new_filter_stronger(operation: str, new_filter_value: Any, old_filter_value: Any) -> bool:
```

Comment:
That type will be inferred from the caller, and in practice will turn out to be `Any`. I don't think mypy is reliable for checks like these.

Reply:
I agree — but I also don't see a reason why not to make explicit the constraint that the docstring promises, even if in practice it ends up not being able to catch anything.

Reply:
Because It's not worth the effort. If I add a `TypeVar` mypy will complain that it doesn't have a `<=` operation. Then I'll have to specify `FieldValueType = TypeVar("FieldValueType", int, str, datetime.date, datetime.datetime)`, but then how do I know it's only one of these types? Because we only do range reasoning on them. Then I'll have to put that in the spec here, and in all the intermediate fnctions. It's just not worth the review cycles. It's self-explanatory that a field we are using the `<=` operator on has `<=` functionality in python.